src/core/thread/network_data_leader.cpp

```diff
@@ -260,12 +260,26 @@ bool LeaderBase::IsOnMesh(const Ip6::Address &aAddress) const
 
     while ((prefix = FindNextMatchingPrefix(aAddress, prefix)) != NULL)
     {
-        if (FindBorderRouter(*prefix) == NULL)
+        // check both stable and full Border Router TLVs
+        for (int i = 0; i < 2; i++)
         {
-            continue;
-        }
+            const BorderRouterTlv *borderRouter = FindBorderRouter(*prefix, /* aStable */ (i == 0));
+
+            if (borderRouter == nullptr)
+            {
+                continue;
+            }
 
-        ExitNow(rval = true);
+            for (uint8_t entry = 0; entry < borderRouter->GetNumEntries(); entry++)
+            {
+                const BorderRouterEntry *borderRouterEntry = borderRouter->GetEntry(entry);
+
+                if (borderRouterEntry->IsOnMesh())
```

Comment:
nit: Works fine as is. Suggestion to use `entry` pointer?
```suggestion
            for (const BorderRouterEntry *entry = borderRouter->GetFirstEntry(); entry <= borderRouter->GetLastEntry();
                 entry                          = entry->GetNext())
            {
                if (entry->IsOnMesh())
```